dev/test/query.ts

```diff
@@ -773,6 +788,135 @@ describe('query interface', () => {
     });
   });
 
+  it('handles stream exception after initialization and heartbeat', () => {
+    const deadlineExceededErr = new GoogleError();
+    deadlineExceededErr.code = Status.DEADLINE_EXCEEDED;
+    deadlineExceededErr.message = 'DEADLINE_EXCEEDED error message';
+
+    let attempts = 0;
+
+    const overrides: ApiOverride = {
+      runQuery: () => {
+        ++attempts;
+
+        // A heartbeat message will initialize the stream, but it is not
+        // a document, so it does not represent progress made on the stream.
+        return stream(heartbeat(1000), deadlineExceededErr);
+      },
+    };
+
+    return createInstance(overrides).then(firestoreInstance => {
+      firestore = firestoreInstance;
+      return firestore
+        .collection('collectionId')
+        .get()
+        .then(() => {
+          throw new Error('Unexpected success in Promise');
+        })
+        .catch(err => {
+          expect(err.message).to.equal('DEADLINE_EXCEEDED error message');
+
+          // The heartbeat initialized the stream before there was a stream
+          // exception, so we only expect a single attempt at streaming.
+          expect(attempts).to.equal(1);
+        });
+    });
+  });
+
+  it('handles retryable exception until progress stops (with get())', () => {
+    let attempts = 0;
+
+    // count of stream initializations that are successful
+    // and make progress (a document is received before error)
+    const initializationsWithProgress = 20;
+
+    // Receiving these error codes on a stream after the stream has received document data
+    // should result in the stream retrying indefinitely.
+    const retryableErrorCodes = [
+      Status.DEADLINE_EXCEEDED,
+      Status.UNAVAILABLE,
+      Status.INTERNAL,
+      Status.UNAVAILABLE,
+    ];
+
+    const overrides: ApiOverride = {
+      runQuery: () => {
```

Comment:
Is it possible to verify the request being sent to `runQuery`. We want to see that `startAfter` is being set correctly. I think this can be derived from `attempts`.

Reply:
Good call, this was something I also wanted to do but forgot.